### PR TITLE
Rudimentary touch support for LUA script

### PR DESCRIPTION
**Pull request by rotorman**
In order to use on radios without physical controls, LUA script would benefit from touch support.
This PR adds rudimentary touch support via swiping and tapping and adds an exit menu to be able to exit the script via touch.

Following video shows how swiping up/down selects items and right/left changes the value. Tapping edits or enters the value:
https://www.youtube.com/watch?v=stvol0fS_pU

### CODE REVIEW

File: src/lua/elrsV2.lua [hunk 1]
```diff
@@ -967,20 +1037,24 @@ local function init()
 end
 
 -- Main
-local function run(event)
+local function run(event, touchState)
```
Comment (pkendall64) [inline src/lua/elrsV2.lua:1040]: Is this backward compatible i.e. will it still work on radios without touch?

File: src/lua/elrsV2.lua [hunk 2]
```diff
@@ -801,23 +813,77 @@ local function handleDevicePageEvent(event)
...
+              incrField(1)
+            elseif touchState.swipeLeft then
+              incrField(-1)
+            elseif touchState.swipeUp then
+            elseif touchState.swipeDown then
+            else
+              -- Sliding but not swiping
+            end
+		  end
+	    else -- event ~= 0 and touchState == nil: key event
```
Comment (pkendall64) [inline src/lua/elrsV2.lua:839]: The indentation is off here

File: src/lua/elrsV2.lua [hunk 3]
```diff
@@ -801,23 +813,77 @@ local function handleDevicePageEvent(event)
...
     end
   else
-    if event == EVT_VIRTUAL_NEXT then
-      selectField(1)
-    elseif event == EVT_VIRTUAL_PREV then
-      selectField(-1)
-    end
+    if event ~= nil then
+      if event ~= 0 then
+	    if touchState then -- Only touch events come with a touchState
```
Comment (pkendall64) [inline src/lua/elrsV2.lua:851]: indentation

File: src/lua/elrsV2.lua [hunk 4]
```diff
@@ -801,23 +813,77 @@ local function handleDevicePageEvent(event)
...
+            elseif touchState.swipeLeft then
+            elseif touchState.swipeUp then
+              selectField(-1)
+            elseif touchState.swipeDown then
+              selectField(1)
+            else
+              -- Sliding but not swiping
+            end
+		      end
+	    else -- event ~= 0 and touchState == nil: key event
```
Comment (pkendall64) [inline src/lua/elrsV2.lua:872]: indentation

File: src/lua/elrsV2.lua [hunk 5]
```diff
@@ -801,23 +813,77 @@ local function handleDevicePageEvent(event)
...
+          if (event == EVT_VIRTUAL_NEXT) then
+            selectField(1)
+          elseif (event == EVT_VIRTUAL_PREV) then
+            selectField(-1)
+          end
+		    end
+      end
+	end
```
Comment (pkendall64) [inline src/lua/elrsV2.lua:880]: indentation

File: src/lua/elrsV2.lua [hunk 6]
```diff
@@ -801,23 +813,77 @@ local function handleDevicePageEvent(event)
...
+    if event ~= nil then
+      if event ~= 0 then
+	if touchState then -- Only touch events come with a touchState
+          if event == EVT_TOUCH_FIRST then
+            -- When the finger first hits the screen
+          elseif event == EVT_TOUCH_BREAK then
+            -- When the finger leaves the screen and did not slide on it
+          elseif event == EVT_TOUCH_TAP then
+            -- A short tap gives TAP instead of BREAK
+            -- touchState.tapCount shows number of taps
```
Comment (pkendall64) [inline src/lua/elrsV2.lua:825]: Can remove these empty cases...

File: src/lua/elrsV2.lua [hunk 7]
```diff
@@ -801,23 +813,77 @@ local function handleDevicePageEvent(event)
...
+          elseif event == EVT_TOUCH_SLIDE then
+            -- Sliding the finger gives a SLIDE instead of BREAK or TAP
+            if touchState.swipeRight then
+              -- Is true if user swiped right
+            elseif touchState.swipeLeft then
```
Comment (pkendall64) [inline src/lua/elrsV2.lua:863]: Can remove these empty cases...

File: src/lua/elrsV2.lua [hunk 8]
```diff
@@ -801,23 +813,77 @@ local function handleDevicePageEvent(event)
...
+            elseif touchState.swipeUp then
+              selectField(-1)
+            elseif touchState.swipeDown then
+              selectField(1)
+            else
+              -- Sliding but not swiping
```
Comment (pkendall64) [inline src/lua/elrsV2.lua:869]: Can remove these empty cases...

File: src/lua/elrsV2.lua (hunk 2 above)
Comment (rotorman) [reply src/lua/elrsV2.lua]: Should be fixed in latest commit

File: src/lua/elrsV2.lua (hunk 3 above)
Comment (rotorman) [reply src/lua/elrsV2.lua]: Should be fixed in latest commit

File: src/lua/elrsV2.lua (hunk 4 above)
Comment (rotorman) [reply src/lua/elrsV2.lua]: Should be fixed in latest commit

File: src/lua/elrsV2.lua (hunk 5 above)
Comment (rotorman) [reply src/lua/elrsV2.lua]: Should be fixed in latest commit

File: src/lua/elrsV2.lua (hunk 6 above)
Comment (rotorman) [reply src/lua/elrsV2.lua]: Unused cases are removed in latest commit

File: src/lua/elrsV2.lua [hunk 9]
```diff
@@ -801,23 +813,77 @@ local function handleDevicePageEvent(event)
...
+            -- Sliding the finger gives a SLIDE instead of BREAK or TAP
+            if touchState.swipeRight then
+              -- Is true if user swiped right
+              incrField(1)
+            elseif touchState.swipeLeft then
+              incrField(-1)
+            elseif touchState.swipeUp then
+            elseif touchState.swipeDown then
+            else
+              -- Sliding but not swiping
```
[the remark this answers is not in the document]
Comment (rotorman) [reply src/lua/elrsV2.lua]: Unused cases are removed in latest commit

File: src/lua/elrsV2.lua [hunk 10]
```diff
@@ -801,23 +813,77 @@ local function handleDevicePageEvent(event)
...
+    if event ~= nil then
+      if event ~= 0 then
+	    if touchState then -- Only touch events come with a touchState
+          if event == EVT_TOUCH_FIRST then
+            -- When the finger first hits the screen
+          elseif event == EVT_TOUCH_BREAK then
+            -- When the finger leaves the screen and did not slide on it
+          elseif event == EVT_TOUCH_TAP then
+            -- A short tap gives TAP instead of BREAK
+            -- touchState.tapCount shows number of taps
```
[the remark this answers is not in the document]
Comment (rotorman) [reply src/lua/elrsV2.lua]: Unused cases are removed in latest commit

File: src/lua/elrsV2.lua (hunk 7 above)
Comment (rotorman) [reply src/lua/elrsV2.lua]: Unused cases are removed in latest commit

File: src/lua/elrsV2.lua (hunk 8 above)
Comment (rotorman) [reply src/lua/elrsV2.lua]: Unused cases are removed in latest commit

File: src/lua/elrsV2.lua (hunk 1 above)
Comment (rotorman) [reply src/lua/elrsV2.lua:1014]: Tested with Zorro running EdgeTX v2.8.0-dev, the LUA runs as well.

File: src/lua/elrsV2.lua [hunk 11]
```diff
@@ -63,6 +65,9 @@ local function allocateFields()
   if folderAccess ~= nil then
     fields[backButtonId].parent = folderAccess
   end
+  fields_count = fields_count + 1
```
[the remark this answers is not in the document]
Comment (rotorman) [reply src/lua/elrsV2.lua]: Fixed, thanks for finding this!

File: src/lua/elrsV2.lua [hunk 12]
```diff
@@ -801,23 +820,51 @@ local function handleDevicePageEvent(event)
...
-      selectField(1)
-    elseif event == EVT_VIRTUAL_PREV then
-      selectField(-1)
+    if event ~= nil then
+      if event ~= 0 then
+        if touchState then
+          if event == EVT_TOUCH_SLIDE then
+            if touchState.swipeLeft then
+              selectField(-1)
+            elseif touchState.swipeRight then
```
Comment (StonedDawg) [inline src/lua/elrsV2.lua:848]: and this to .swipeUp? so swipe up/down is acting like the scroll behaviour

File: src/lua/elrsV2.lua [hunk 13]
```diff
@@ -801,23 +820,51 @@ local function handleDevicePageEvent(event)
...
   else
-    if event == EVT_VIRTUAL_NEXT then
-      selectField(1)
-    elseif event == EVT_VIRTUAL_PREV then
-      selectField(-1)
+    if event ~= nil then
+      if event ~= 0 then
+        if touchState then
+          if event == EVT_TOUCH_SLIDE then
+            if touchState.swipeLeft then
```
[the remark this answers is not in the document]
Comment (rotorman) [reply src/lua/elrsV2.lua]: Please have a look at [905f096](https://github.com/ExpressLRS/ExpressLRS/pull/1515/commits/905f096594f43acd014a0fac0a0a50b65665b219)

File: src/lua/elrsV2.lua (hunk 12 above)
Comment (rotorman) [reply src/lua/elrsV2.lua]: Please have a look at [905f096](https://github.com/ExpressLRS/ExpressLRS/pull/1515/commits/905f096594f43acd014a0fac0a0a50b65665b219)

**Review by StonedDawg — APPROVED** [review]
this is working great on ETX and OTX radio...but older freedomtx is unknown

**Comment by pkendall64** [comment]
I tested this as well. Works on on the TX16S, but I'd still use the scroll wheel as just swiping to do things just feels odd.

**Review by pkendall64 — APPROVED** [review]

**Review by wvarty — APPROVED** [review]
Tested on Tango2. Loaded without error and operates as expected.

**Pull request merged**